test/test_av.py

```diff
@@ -0,0 +1,265 @@
+    unit_nodes = group.unit_nodes
+    modes = group.mode_ids()
+    order = group.order
+
+    # loop over modes and check smoothness
+    for i, mode in enumerate(modes):
+        ele_soln = basis[i](unit_nodes)
+        soln[0].set(np.tile(ele_soln, (nele, 1)))
+        if sum(mode) == order:
+            expected = 1.0
+        else:
+            expected = 0.0
+        err = norm_indicator(expected, discr, soln)
+        assert err < tolerance,  "Only highest modes should trigger indicator (1.0)"
+
+    # Test s0
+    s0 = -1.
+    eps = 1.0e-6
+
+    phi_n_p = np.sqrt(np.power(10, s0))
+    phi_n_pm1 = np.sqrt(1. - np.power(10, s0))
+
+    # pick a polynomial of order n_p, n_p-1
+    n_p = np.array(np.nonzero((np.sum(modes, axis=1) == order))).flat[0]
+    n_pm1 = np.array(np.nonzero((np.sum(modes, axis=1) == order-1))).flat[0]
+
+    # create test soln perturbed around
+    # Solution above s0
+    ele_soln = ((phi_n_p+eps)*basis[n_p](unit_nodes)
+                + phi_n_pm1*basis[n_pm1](unit_nodes))
+    soln[0].set(np.tile(ele_soln, (nele, 1)))
+    err = norm_indicator(1.0, discr, soln, s0=s0, kappa=0.0)
+    assert err < tolerance,  (
+        "A function with an indicator >s0 should trigger indicator")
+
+    # Solution below s0
+    ele_soln = ((phi_n_p-eps)*basis[n_p](unit_nodes)
+                + phi_n_pm1*basis[n_pm1](unit_nodes))
+    soln[0].set(np.tile(ele_soln, (nele, 1)))
+    err = norm_indicator(0.0, discr, soln, s0=s0, kappa=0.0)
+    assert err < tolerance, (
+        "A function with an indicator <s0 should not trigger indicator")
+
+    # Test kappa
+    # non-perturbed solution
+    # test middle value
+    kappa = 0.5
+    ele_soln = (phi_n_p*basis[n_p](unit_nodes)
+                + phi_n_pm1*basis[n_pm1](unit_nodes))
+    soln[0].set(np.tile(ele_soln, (nele, 1)))
+    err = norm_indicator(0.5, discr, soln, s0=s0, kappa=kappa)
+    assert err < 1.0e-10,  "A function with s_e=s_0 should return 0.5"
+
+    # test bounds
+    # lower bound
+    shift = 1.0e-5
+    err = norm_indicator(0.0, discr, soln, s0=s0+kappa+shift, kappa=kappa)
+    assert err < tolerance,  "s_e<s_0-kappa should not trigger indicator"
+    err = norm_indicator(0.0, discr, soln, s0=s0+kappa-shift, kappa=kappa)
+    assert err > tolerance,  "s_e>s_0-kappa should trigger indicator"
+
+    # lower bound
+    err = norm_indicator(1.0, discr, soln, s0=s0-(kappa+shift), kappa=kappa)
+    assert err < tolerance,  "s_e>s_0+kappa should fully trigger indicator (1.0)"
+    err = norm_indicator(1.0, discr, soln, s0=s0-(kappa-shift), kappa=kappa)
+    assert err > tolerance,  "s_e<s_0+kappa should not fully trigger indicator (1.0)"
+
+
+@pytest.mark.parametrize("dim",  [1, 2, 3])
+@pytest.mark.parametrize("order",  [2, 3])
+def test_artificial_viscosity(ctx_factory, dim, order):
+    """Test artificial_viscosity.
+
+    Tests the application on a few simple functions
+    to confirm artificial viscosity returns the analytical result.
+    """
```

Comment:
I think this should be replaced by a manufactured-solution test, instead of these two simplistic known-solution tests.

Reply:
I tried quickly re-use the diffusion operator MMS tests, but turns out not to be as easy of a drop-in as I had imagined.

We'll need some infrastructure backing to make this happen.   It looks like the amount of effort to get an MMS infrastructure for AV is about the same as replacing the guts of AV with the diffusion operator.  I think we'd prefer using the diffusion operator.   Another option would be just beefing up the tests a bit with more analytic functions, perhaps leveraging the symbolic infrastructure.

Reply:
@lukeolson added a trig test that helps with this a little.

I tried to add a discontinuous coeff test (which I do not expect this operator to pass) - but it turned out to be too much effort for this fluid-specific operator.   If this operator was just a (multi)scalar laplacian operator, then it'd be a no-brainer.  Alas, now it re-uses the gradient operator and boundary projection routines designed specifically for fluid solutions - and thus requires physically-consistent fluid solutions.